### ISSUE TITLE
Add Seperators For Foreach Question Block

### ISSUE THREAD

**Issue by Bennett-Woolf** [maths/moodle-qtype_stack#1020]
Hi all,

I find when I am using [[foreach]] question blocks, I often am using it to shortcut writing out a list of variables. As such, I want to put commas and a space between each run of the foreach loop except the last one, as I don't want an extra comma and space at the end. My solution so far is to either not use the forloop (if there aren't very many elements), or drop out of it one element early so that I can add in the special case manually.

Would it be possible to either add a "seperator" argument that inserts whatever you input between each run of the forloop?

The most recent example of this phenomena is "[[foreach j="makelist(i,i,1,length(complexNumbers))"]]  z_{@j@} = {@complexNumbers[j]@}, \quad [[/ foreach]]", which, for my case, would output

z_1 = x_1 + i y_1,     z_2 = x_2 + i y_2,     z_3 = x_3 + y_3,     (This is where the quad ends)

I hope others would find this useful too, or if anyone has a workaround, it would be much appreciated,

Jake

**Comment by aharjula** [comment]
#917 might have an interesting use of `[[define]]` in one of the suggestions that might work for this as well. But yes someone could possibly add such an extra separator argument to the block. It would only reserve one identifier for something that could not be used as a name of an iteration variable. Not promising to add such a thing though, maybe some newer coder might want to wrap their head around this and add such logic to [this](https://github.com/maths/moodle-qtype_stack/blob/master/stack/cas/castext2/blocks/foreach.block.php). We might at the same time consider if we need to bring other common loop concepts, like loop local flags for first and last executions etc. into the block.

Basically, I would write your example like this, in a world where we would not have such a separator option:
```
[[define ComplexNumbers="[1,1-i,2+i,2*i]"/]]
[[define _first="true"/]][[foreach j="makelist(i,i,1,length(ComplexNumbers))"]][[if test="_first"]][[define _first="false"/]][[else]], \quad  [[/if]]z_{@j@} = {@ComplexNumbers[j]@}[[/ foreach]]
```
*\quad might not work there though.*

Basically, `[[define/]]` allows one to change the values of variables and define new ones within the text, including in the conditional and loopy bits...

**Comment by Bennett-Woolf** [comment]
Hi Matti,

I've been thinking about your response (and had the situation happen again) and a simpler solution to my problem is do exactly what you're saying but without the define blocks and just manually unroll the first loop, giving us

```
[[define ComplexNumbers="[1,1-i,2+i,2*i]"/]]
[[define loopList="makelist(i,i,1,length(ComplexNumbers))"/]]
z_{@first(loopList)@}[[foreach j="rest(loopList)"]], \quad  z_{@j@} = {@ComplexNumbers[j]@}[[/ foreach]]
```

which avoids having an extra conditional. The idea of unrolling the first rather than the last item in the loop is exactly what I needed so thank you so much for that.

Unfortunately my PHP knowledge is completely non-existent, otherwise I would love to add foreach constructs myself.

Thank you for all the help

**Comment by sangwinc** [comment]
This is clearly a needed feature.  We can accommodate "implode" within the current setup.  So, the question now is do we need "sugar" to make this really easy (and clear) or should we just use the suggestions above, and the documentation example I've just pushed?

**Comment by aharjula** [comment]
If we are now resurrecting this, there is a thing to note. While back in 2023 I said: *"...yes someone could possibly add such an extra separator argument to the block. It would only reserve one identifier for something that could not be used as a name of an iteration variable."*, I am starting to wonder how such an argument should be named in a world where people have been using `[[foreach]]` for quite some time (some since STACK 3 (abacus branch) most from 4) and might have already used all the sensible identifiers. I think that even I have at some point interleaved something where one was logically called a separator (even if it varied and came from a list).

If we start reserving some identifiers, then we need to think about possible collisions and balance that with usability. `sep` or `separator` are obviously ideal for usability and understandability, but they also have a reasonable probability for collisions. A prefix of some kind might lessen the collisions, but would make the author write more and more, and more importantly, nonsense from their point of view.

I really should not have made `[[foreach]]` to consider all parameters as variables; had I thought things through, I should have reserved some words or at least a prefix, or something like `loop_options`. This might be an issue elsewhere in STACK as well. Do we even have any reserved STACK internals-only namespace anywhere other than the ones enforced by the security-map?

The thing is that if we start adding reserved words here, we might as well add multiple at the same time:
 - Something for that separator.
 - Something to signal the start of the loop.
 - Something to signal the end of the loop.
 - Something to store the index.
 - Similar signals and counters for ancestor loops in a stack of some kind.

The problem is that while the first is obviously something going in as a value to a fixed receiving identifier, the following ones are something that would actually also affect those identifiers in the body of the loop. I would be somewhat disappointed if I managed to mess up people's core vocabulary after I spent so much time getting the word `system` to be acceptable. It is already annoying that one cannot use `labels` when building tables or charts, as that is a keyword...

**Comment by sangwinc** [comment]
This is a classic dilemma @aharjula: do we have an argument to an existing function (`separator` option to the `[[foreach ...]]` block) or a new top-level block (e.g. `[[implode]]`)?  I don't have strong preferences here, but it's interesting other people really do make use of this functionality.  We could just stay with the documentation I pushed today which allows authors to solve the problem!
